crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -185,56 +265,1249 @@ impl<S: States> NextState<S> {
+                      current_state: Option<ResMut<State<T>>>,
+                      state_set: Option<Res<State<S::RawState>>>| {
+            if parent_changed.is_empty() {
+                return;
             }
+            parent_changed.clear();
+
+            let new_state =
+                if let Some(state_set) = S::convert_to_usable_state(state_set.as_deref()) {
+                    T::compute(state_set)
+                } else {
+                    None
+                };
+
+            internal_apply_state_transition(event, commands, current_state, new_state);
         };
+
+        schedule
+            .add_systems(system.in_set(ApplyStateTransition::<T>::apply()))
+            .add_systems(
+                should_run_transition::<T, OnEnter<T>>
+                    .pipe(run_enter::<T>)
+                    .in_set(StateTransitionSteps::EnterSchedules),
+            )
+            .add_systems(
+                should_run_transition::<T, OnExit<T>>
+                    .pipe(run_exit::<T>)
+                    .in_set(StateTransitionSteps::ExitSchedules),
+            )
+            .add_systems(
+                should_run_transition::<T, OnTransition<T>>
+                    .pipe(run_transition::<T>)
+                    .in_set(StateTransitionSteps::TransitionSchedules),
+            )
+            .configure_sets(
+                ApplyStateTransition::<T>::apply()
+                    .in_set(StateTransitionSteps::DependentTransitions)
+                    .after(ApplyStateTransition::<S::RawState>::apply()),
+            );
+    }
+
+    fn register_sub_state_systems_in_schedule<T: SubStates<SourceStates = Self>>(
+        schedule: &mut Schedule,
+    ) {
+        let system = |mut parent_changed: EventReader<StateTransitionEvent<S::RawState>>,
+                      event: EventWriter<StateTransitionEvent<T>>,
+                      commands: Commands,
+                      current_state: Option<ResMut<State<T>>>,
+                      state_set: Option<Res<State<S::RawState>>>| {
+            if parent_changed.is_empty() {
+                return;
+            }
+            parent_changed.clear();
+
+            let new_state =
+                if let Some(state_set) = S::convert_to_usable_state(state_set.as_deref()) {
+                    T::should_exist(state_set)
+                } else {
+                    None
+                };
+
+            match new_state {
+                Some(value) => {
+                    if current_state.is_none() {
+                        internal_apply_state_transition(
+                            event,
+                            commands,
+                            current_state,
+                            Some(value),
+                        );
+                    }
+                }
+                None => {
+                    internal_apply_state_transition(event, commands, current_state, None);
+                }
+            };
```

Comment:
So this only runs transitions when sub-state goes in and out of existence, and the rest is handled by `apply_state_transitions`.
Is that correct?

Reply:
Yes - should_exist only sets the state if/when it goes in and out of existance.